diff --git a/kern/init.c b/kern/init.c
index 27e6ab5..3d94128 100644
--- a/kern/init.c
+++ b/kern/init.c
@@ -71,9 +71,9 @@ i386_init(void)
 #else
 	// Touch all you want.
 	// ENV_CREATE(user_primes, ENV_TYPE_USER);
-	ENV_CREATE(user_yield, ENV_TYPE_USER);
-	ENV_CREATE(user_yield, ENV_TYPE_USER);
-	ENV_CREATE(user_yield, ENV_TYPE_USER);
+	ENV_CREATE(user_faultalloc, ENV_TYPE_USER);
+	//ENV_CREATE(user_yield, ENV_TYPE_USER);
+	//ENV_CREATE(user_yield, ENV_TYPE_USER);
 #endif // TEST*
 
 	// Schedule and run the first user environment!
diff --git a/kern/syscall.c b/kern/syscall.c
index d422846..3f6b3be 100644
--- a/kern/syscall.c
+++ b/kern/syscall.c
@@ -455,6 +455,8 @@ syscall(uint32_t syscallno, uint32_t a1, uint32_t a2, uint32_t a3, uint32_t a4,
 		return sys_page_map(a1, (void*)a2, a3, (void *)a4, a5);
 	case SYS_page_unmap:
 		return sys_page_unmap(a1, (void*)a2);
+	case SYS_env_set_pgfault_upcall:
+		return sys_env_set_pgfault_upcall((envid_t)a1, (void *)a2);
 	default:
 		return -E_INVAL;
 	}
diff --git a/kern/trap.c b/kern/trap.c
index 360bca1..60a36d1 100644
--- a/kern/trap.c
+++ b/kern/trap.c
@@ -443,16 +443,16 @@ page_fault_handler(struct Trapframe *tf)
 	// UXSTACKTOP), then branch to curenv->env_pgfault_upcall.
 	// Call the environment's page fault upcall
 	if(curenv->env_pgfault_upcall){
-		struct UTrapframe *utf = NULL;;
+		struct UTrapframe utf;
 
 		// TODO: JULIANNE 
 		// Fill in UTrapframe:
-		utf->utf_fault_va = fault_va;
-		utf->utf_err = tf->tf_err;
-		utf->utf_regs = tf->tf_regs;
-		utf->utf_eip = tf->tf_eip;
-		utf->utf_eflags = tf->tf_eflags;
-		utf->utf_esp = tf->tf_esp;
+		utf.utf_fault_va = fault_va;
+		utf.utf_err = tf->tf_err;
+		utf.utf_regs = tf->tf_regs;
+		utf.utf_eip = tf->tf_eip;
+		utf.utf_eflags = tf->tf_eflags;
+		utf.utf_esp = tf->tf_esp;
 
 
 		// Check if already on the exception stack
@@ -472,7 +472,7 @@ page_fault_handler(struct Trapframe *tf)
 
 		// Modify env's trapframe: return upcall handler
 		tf->tf_eip = (uintptr_t)curenv->env_pgfault_upcall;
-		tf->tf_esp = (uintptr_t)utf;
+		*((struct UTrapframe *) tf->tf_esp) = utf;
 
 		// user_mem_assert(curenv, (uintptr_t)tf->tf_esp, sizeof(struct UTrapframe), PTE_W | PTE_P | PTE_U);
 		user_mem_assert(curenv, (void*)tf->tf_esp, sizeof(struct UTrapframe), PTE_P| PTE_W | PTE_U);
